compliance-monitor/monitor.py

```diff
@@ -0,0 +1,525 @@
+    yaml = ruamel.yaml.YAML(typ='safe')
+    with open(yaml_path, "r") as fileobj:
+        document = yaml.load(fileobj.read())
+    # The following will also delete entries that are not present in the given yaml.
+    # It is paramount that all extant primary keys be kept because the reports refer to them, and
+    # deletions will cascade! But we should only ever delete entries for non-stable versions; stable versions
+    # are deemed immutable except maybe for checks, and for those, at least the ids are immutable.
+    with conn.cursor() as cur:
+        scopeid = db_update_scope(cur, document)
+        all_versions = set()
+        for vdata in document['versions']:
+            versionid = db_update_version(cur, scopeid, vdata)
+            all_versions.add(versionid)
+            all_standards = set()
+            for sdata in vdata['standards']:
+                standardid = db_update_standard(cur, versionid, sdata)
+                all_standards.add(standardid)
+                all_checks = set()
+                for cdata in sdata.get('checks', ()):
+                    checkid = db_update_check(cur, versionid, standardid, cdata)
+                    all_checks.add(checkid)
+                db_filter_checks(cur, standardid, lambda checkid, *_: checkid in all_checks)
+            db_filter_standards(cur, versionid, lambda standardid, *_: standardid in all_standards)
+        db_filter_versions(cur, scopeid, lambda versionid, *_: versionid in all_versions)
+        conn.commit()
+
+
+def import_cert_yaml_dir(yaml_path, conn):
+    for fn in sorted(os.listdir(yaml_path)):
+        if fn.startswith('scs-') and fn.endswith('.yaml'):
+            import_cert_yaml(os.path.join(yaml_path, fn), conn)
+
+
+def import_templates(template_dir, env, templates):
+    for fn in os.listdir(template_dir):
+        if fn.startswith("."):
+            continue
+        name = fn  # used to be fn.rsplit(".", 1)[0]
+        if name not in templates:
+            continue
+        with open(os.path.join(template_dir, fn), "r") as fileobj:
+            templates[name] = env.from_string(fileobj.read())
+
+
+def validate_templates(templates, required_templates=REQUIRED_TEMPLATES):
+    missing = [key for key in required_templates if not templates.get(key)]
+    if missing:
+        raise RuntimeError(f"missing templates: {', '.join(missing)}")
+
+
+async def auth(request: Request, conn: Annotated[connection, Depends(get_conn)]):
+    return get_current_account(await security(request), conn)
+
+
+async def optional_auth(request: Request, conn: Annotated[connection, Depends(get_conn)]):
+    return get_current_account(await optional_security(request), conn)
+
+
+def check_role(account: Optional[tuple[str, str]], subject: str = None, roles: int = 0):
+    if account is None:
+        raise HTTPException(status_code=401, detail="Permission denied")
+    current_subject, present_roles = account
+    if subject != current_subject and roles & present_roles != roles:
```

Comment:
To be clear: the `and` is intended? (Account `xyz` can always see own data, independent of roles. If so, this intent could be added as a doc comment.)

Reply:
Yes, an account can always access its own data. I can add a comment for sure.